modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/ScaledFloatFieldMapper.java

```diff
@@ -59,124 +54,82 @@
 import java.time.ZoneId;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 
 /** A {@link FieldMapper} for scaled floats. Values are internally multiplied
  *  by a scaling factor and rounded to the closest long. */
-public class ScaledFloatFieldMapper extends FieldMapper {
+public class ScaledFloatFieldMapper extends ParametrizedFieldMapper {
 
     public static final String CONTENT_TYPE = "scaled_float";
+
     // use the same default as numbers
     private static final Setting<Boolean> COERCE_SETTING = NumberFieldMapper.COERCE_SETTING;
-    private static final FieldType FIELD_TYPE = new FieldType();
-    static {
-        FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+
+    private static ScaledFloatFieldMapper toType(FieldMapper in) {
+        return (ScaledFloatFieldMapper) in;
     }
 
-    public static class Builder extends FieldMapper.Builder<Builder> {
+    public static class Builder extends ParametrizedFieldMapper.Builder {
 
-        private boolean scalingFactorSet = false;
-        private double scalingFactor;
-        private Boolean ignoreMalformed;
-        private Boolean coerce;
-        private Double nullValue;
+        private final Parameter<Boolean> indexed = Parameter.indexParam(m -> toType(m).indexed, true);
+        private final Parameter<Boolean> hasDocValues = Parameter.docValuesParam(m -> toType(m).hasDocValues, true);
+        private final Parameter<Boolean> stored = Parameter.storeParam(m -> toType(m).stored, false);
 
-        public Builder(String name) {
-            super(name, FIELD_TYPE);
-            builder = this;
-        }
+        private final Parameter<Explicit<Boolean>> ignoreMalformed;
+        private final Parameter<Explicit<Boolean>> coerce;
 
-        public Builder ignoreMalformed(boolean ignoreMalformed) {
-            this.ignoreMalformed = ignoreMalformed;
-            return builder;
-        }
+        private final Parameter<Double> scalingFactor = new Parameter<>("scaling_factor", false, () -> null,
+            (n, c, o) -> XContentMapValues.nodeDoubleValue(o), m -> toType(m).scalingFactor)
+            .setValidator(v -> {
+                if (v == null) {
+                    throw new IllegalArgumentException("Field [scaling_factor] is required");
+                }
+                if (Double.isFinite(v) == false || v <= 0) {
+                    throw new IllegalArgumentException("[scaling_factor] must be a positive number, got [" + v + "]");
+                }
+            });
+        private final Parameter<Double> nullValue = new Parameter<>("null_value", false, () -> null,
```

Comment:
Annoyingly, we can't because we need to take `null` as a default value in order to check that it has been set, and `Parameter.doubleParameter` doesn't allow you to do that.